`#[cfg]` attribute is not handled properly on struct fields

```rust
pub struct ReadDir {
    pub inner: i32,
    #[cfg(not(A))]
    pub end_of_stream: bool,
    #[cfg(A)]
    pub end_of_stream_but_different: bool,
}

fn main() {
    let _ = ReadDir {
        inner: 14,
        end_of_stream: false,
        end_of_stream_but_different: false,
    };
}
```

slightly modified version of an exemple from `std` 1.49. This code should error out in all cases as both `cfg` directives are contradictory, so at least one field should be nonexistent in the instantiation. `rustc` outputs the following with no particular cfg command line options:

```rust
error[E0560]: struct `ReadDir` has no field named `end_of_stream_but_different`
  --> test.rs:13:9
   |
13 |         end_of_stream_but_different: false,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `ReadDir` does not have this field
   |
   = note: all struct fields are already assigned

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0560`.
```

but `gccrs` compiles the file happily

--- a/gcc/rust/expand/rust-cfg-strip.cc
+++ b/gcc/rust/expand/rust-cfg-strip.cc
@@ -195,6 +195,26 @@ CfgStrip::maybe_strip_struct_fields (std::vector<AST::StructField> &fields)
 }
 
 void
+CfgStrip::maybe_strip_struct_expr_fields (
+  std::vector<std::unique_ptr<AST::StructExprField>> &fields)
+{
+  for (auto it = fields.begin (); it != fields.end ();)
+    {
+      auto &field = *it;
+
+      auto &field_attrs = field->get_outer_attrs ();
+      expand_cfg_attrs (field_attrs);
+      if (fails_cfg_with_expand (field_attrs))
+	{
+	  it = fields.erase (it);
+	  continue;
+	}
+
+      ++it;
+    }
+}
+
+void
 CfgStrip::maybe_strip_tuple_fields (std::vector<AST::TupleField> &fields)
 {
   for (auto it = fields.begin (); it != fields.end ();)
@@ -962,6 +982,8 @@ CfgStrip::visit (AST::StructExprStructFields &expr)
 		       "cannot strip expression in this position - outer "
 		       "attributes not allowed");
     }
+
+  maybe_strip_struct_expr_fields (expr.get_fields ());
 }
 
 void
@@ -1852,6 +1874,10 @@ CfgStrip::visit (AST::StructStruct &struct_item)
     }
 
   AST::DefaultASTVisitor::visit (struct_item);
+
+  /* strip struct fields if required - this is presumably
+   * allowed by spec */
+  maybe_strip_struct_fields (struct_item.get_fields ());
 }
 void
 CfgStrip::visit (AST::TupleStruct &tuple_struct)

--- a/gcc/rust/expand/rust-cfg-strip.h
+++ b/gcc/rust/expand/rust-cfg-strip.h
@@ -36,6 +36,8 @@ public:
   void go (AST::Crate &crate);
 
   void maybe_strip_struct_fields (std::vector<AST::StructField> &fields);
+  void maybe_strip_struct_expr_fields (
+    std::vector<std::unique_ptr<AST::StructExprField>> &fields);
   void maybe_strip_tuple_fields (std::vector<AST::TupleField> &fields);
   void maybe_strip_function_params (
     std::vector<std::unique_ptr<AST::Param>> &params);

--- a/gcc/testsuite/rust/compile/macro-issue2983_2984.rs
+++ b/gcc/testsuite/rust/compile/macro-issue2983_2984.rs
@@ -1,0 +1,27 @@
+pub struct ReadDir {
+    pub inner: i32,
+    #[cfg(not(A))]
+    pub end_of_stream: bool,
+    #[cfg(A)]
+    pub end_of_stream_but_different: bool,
+}
+
+fn main() {
+    // Success
+    let _ = ReadDir {
+        inner: 14,
+        #[cfg(not(A))]
+        end_of_stream: false,
+        #[cfg(A)]
+        end_of_stream_but_different: false,
+    };
+
+    // Error
+    let _ = ReadDir {
+        inner: 14,
+        end_of_stream: false,
+        end_of_stream_but_different: false, // { dg-error "failed to resolve type for field" }
+        // { dg-error "unknown field" "" { target *-*-* } .-1 }
+        // { dg-prune-output "compilation terminated" }
+    };
+}

